odoo/tools/profiler.py

```diff
@@ -1,197 +1,326 @@
 # -*- coding: utf-8 -*-
-from decorator import decorator
-import inspect
+import datetime
+import gc
+import json
 import logging
+import os
 import sys
 import time
+import threading
 
-import odoo
 _logger = logging.getLogger(__name__)
 
 
-class _LogTracer(object):
-    def __init__(self, whitelist=None, blacklist=None, files=None, deep=False):
-        self.profiles = {}
-        self.whitelist = whitelist
-        self.blacklist = blacklist
-        self.files = files
-        self.deep = deep
-        self.first_frame = None
+def _format_frame(frame):
+    co = frame.f_code
+    return (co.co_filename, frame.f_lineno, co.co_name, '')
 
-    def tracer(self, frame, event, arg):
-        if not self.first_frame:
-            self.first_frame = frame.f_code
-        if not self.deep and self.first_frame != frame.f_code:
-            return self.tracer
 
-        if frame.f_code.co_name in ['<genexpr>', '__getattr__', '__iter__', '__init__']:
-            return
+def _format_stack(stack):
+    return [list(frame) for frame in stack]
+
+
+def get_current_frame(thread=None):
+    if thread:
+        frame = sys._current_frames()[thread.ident]
+    else:
+        frame = sys._getframe()
+    while frame.f_code.co_filename == __file__:
+        frame = frame.f_back
+    return frame
+
+
+def _get_stack_trace(frame=None, limit_frame=None):
+    stack = []
+    while frame is not None and frame != limit_frame:
+        stack.append(_format_frame(frame))
+        frame = frame.f_back
+        if limit_frame and frame is None:
+            _logger.error("Limit frame was not found")
+    return list(reversed(stack))
+
+
+def stack_size():
+    frame = sys._getframe().f_back
+    s = 0
+    while frame:
+        s += 1
+        frame = frame.f_back
+    return s
+
+
+def make_session(name=''):
+    return '%s %s' % (datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S'), name)
+
+class Collector():
+    """
+    Generic collector to be inherited.
+    Defines default behaviours to create an entry and output result to database or a file
+    """
+    def __init__(self):
+        self.result = []
+        self.profiler = None
+
+    def _get_stack_trace(self, frame=None):
+        frame = frame or get_current_frame(self.profiler.init_thread)
+        return _get_stack_trace(frame, self.profiler.init_frame)
+
+    def _make_entry(self, frame=None):
+        t0 = time.time()
+        entry = {
+            'stack': self._get_stack_trace(frame),
+            'exec_context': dict(getattr(self.profiler.init_thread, 'exec_context', {})),
```

Comment:
micro-optimization (`dict` takes any iterable):
```suggestion
            'exec_context': dict(getattr(self.profiler.init_thread, 'exec_context', ())),
```